File: ch 5 DFS&BFS/백준/1012.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10000)


def dfs(x, y, n, m):
  if x < 0 or y < 0 or x >= n or y >= m:
    return False
  if graph[x][y] == 1:
    graph[x][y] = 0
    # 상하좌우의 위치도 모두 재귀적으로 호출
    #  재귀 호출 순서는 탐색 순서에만 영향을 미치며, 전체 연결된 컴포넌트를 탐색하는 결과에는 영향을 미치지 X
    dfs(x - 1, y, n, m)
    dfs(x + 1, y, n, m)
    dfs(x, y - 1, n, m)
    dfs(x, y + 1, n, m)
    return True
  return False


t = int(input().strip())

for _ in range(t):
  result = 0
  m, n, k = map(int, input().strip().split())
  graph = [[0] * m for _ in range(n)]  # n이 세로, m이 가로
  #[0] * m 이 길이가 m인 리스트를 만들고 모두 0으로 초기화
  for _ in range(k):
    x, y = map(int, input().strip().split())
    graph[y][x] = 1  # 입력 좌표는 (가로, 세로) 순서로 주어짐

  for i in range(n):
    for j in range(m):
      if dfs(i, j, n, m): # True 일때만 카운트
        result += 1

  print(result)

File: ch 5 DFS&BFS/백준/1012.js
const fs = require("fs");
const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "ex.txt")
  .toString()
  .trim()
  .split("\n");

// 첫 줄: 테스트 케이스의 수
const T = parseInt(input[0]);

// 테스트 케이스 처리
let index = 1; // 현재 처리 중인 줄의 인덱스

for (let t = 0; t < T; t++) {
  let cnt = 0;
  // 각 테스트 케이스의 첫 줄
  const [M, N, K] = input[index++].split(" ").map(Number);

  // 배추밭 초기화 (N x M 크기 2차원 배열)
  const graph = Array.from({ length: N }, () => Array(M).fill(0));

  // 배추 위치 정보 저장
  for (let k = 0; k < K; k++) {
    const [X, Y] = input[index++].split(" ").map(Number);
    graph[Y][X] = 1; // X, Y 좌표에 배추 심기
  }
  const d = [
    [0, 1],
    [0, -1],
    [-1, 0],
    [1, 0],
  ];

  function dfs(y, x) {
    if (x < 0 || x >= M || y < 0 || y >= N) return false;
    if (graph[y][x] === 1) {
      graph[y][x] = 0;
      // 4방향 탐색
      for (const [dy, dx] of d) {
        dfs(y + dy, x + dx);
      }
      return true;
    }
    return false;
  }

  for (let i = 0; i < N; i++) {
    // i가 행
    for (let j = 0; j < M; j++) {
      // j가 열
      if (dfs(i, j)) {
        cnt++;
      }
    }
  }

  console.log(cnt);
}
